### DISCUSSION TITLE
What is the difference between wasm-as and wat2wasm?

### DISCUSSION THREAD

**Discussion by Evian-Zhang** [WebAssembly/binaryen#5435] [Q&A] [answered]
`wasm-as` is a tool in binaryen, and from README, its purpose is

> wasm-as: Assembles WebAssembly in text format (currently S-Expression format) into binary format (going through Binaryen IR).

`wat2wasm` is a tool in wabt, and from README, its purpose is

> [wat2wasm](https://webassembly.github.io/wabt/doc/wat2wasm.1.html): translate from [WebAssembly text format](https://webassembly.github.io/spec/core/text/index.html) to the [WebAssembly binary format](https://webassembly.github.io/spec/core/binary/index.html)

In my understanding, both of tools translates text-format wat to binary-format wasm, and `wasm-as` will additionally experience a binaryen IR stage.

So in view of users, what is the difference between these two tools? And if I want to translates a hand-written wat file to wasm, what is the recommended tool to use? (MDN [uses wabt toolchains](https://developer.mozilla.org/en-US/docs/WebAssembly/Text_format_to_wasm#converting_the_text_.wat_into_a_binary_.wasm_file))

**Comment by tlively** [comment, marked as answer]
WABT's `wat2wasm` translates directly to Wasm without modifying anything and supports the full standard text format. Binaryen's `wasm-as` does not support the full standard text format (although we are working on fixing that) and may modify instruction sequences that are not directly representable in Binaryen IR. If all you want to do is translate wat to Wasm, you're probably better off using `wat2wasm` as long as WABT supports all the Wasm features you're using. On the other hand, if you're developing a toolchain or optimization pipeline that uses Binaryen for other things, you might as well use `wasm-as`.

**Reply by gertvv** [reply]
Is the work to support the full standard text format progressing? As it stands some fairly basic examples from MDN don't compile using `wasm-as` (e.g. their examples on the if statement) and that's quite confusing. I've tried `wat2wasm` but their support for reference types and GC does not meet my needs.

**Reply by tlively** [reply]
The work to switch to parsing the standard text format is complete and is included in the latest release (version 118 at time of writing).

**Reply by gertvv** [reply]
Ah, I recently installed 117 and didn't realize 118 was out. Can confirm the examples compile now. Massive improvement, thanks!
